docs/lliurex-store-cli.1
.TH "LLIUREX\-STORE\-CLI" "1" "November 2018" "" ""
.SH "NAME"
\fBlliurex-store-cli\fR
.SH SYNOPSYS
.P
.RS 2
.nf
lliurex\-store\-cli [OPTIONS] package_name [\-\-appimage] [\-\-snap]
.fi
.RE
.SH DESCRIPTION
.P
.RS 2
.nf
Manage packages from the Lliurex Store\.

Mandatory arguments to long options are mandatory for short options too\.

\-s, \-\-search [name]
    Search *name* in the Lliurex Store\. Name could be a package name, a descriptive string or anything else related with a program\.

\-v, \-\-view [package]
    Shows extended information for package\. Package must be a full package name\.

\-i, \-\-install [package]
    Installs package\. Package must be a full package name\.

\-r, \-\-remove [package]
    Removes package\. Package must be a full package name\.

\-\-appimage
    Enables appimage plugin

\-\-snap
    Enables snap plugin

In order to work with appimages or snaps is mandatory to enable their respective plugins\. Install, view and remove options needed the full package name as returned by the search\. Search can search by name or metainformation of a package thus you can search for "photo" to retrieve all the available applications related to photography (image processing, image cataloguing\.\.\.etc\.\.)
.fi
.RE
.SH EXAMPLES
.P
.RS 2
.nf
* Search for a multimedia player:
lliurex\-store\-cli \-s player

* Search for a CAD application in all sources:
lliurex\-store\-cli \-s CAD \-\-appimage \-\-snap

* View information for package firefox\-esr:
lliurex\-store\-cli \-v firefox\-esr

* View information for appimage package subsurface\.appimage:
lliurex\-store\-cli \-v subsurface\.appimage \-\-appimage`

* Install firefox from snap:
lliurex\-store\-cli \-i firefox\.snap \-\-snap

* Remove epiphany\-browser:
lliurex\-store\-cli \-r epiphany\-browser

* Remove subsurface appimage:
lliurex\-store\-cli \-r subsurface\.appimage \-\-appimage
.fi
.RE
.SH REPORTING BUGS
.P
.RS 2
.nf
Lliurex Github: https://github\.com/lliurex/lliurex\-store/issues
.fi
.RE
.SH SEE ALSO
.P
.RS 2
.nf
Snapcraft: https://snapcraft\.io
Appimage: https://appimage\.org
.fi
.RE



debian/changelog
lliurex-store (0.7) bionic; urgency=medium

  * Implemented cache for package info

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Fri, 02 Nov 2018 09:15:13 +0100

lliurex-store (0.6.2) bionic; urgency=medium

  * Improved icons cache

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 24 Oct 2018 09:49:26 +0200

lliurex-store (0.6.1) bionic; urgency=medium

  * DebManager: add threads

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 23 Oct 2018 09:16:05 +0200

lliurex-store (0.6) bionic; urgency=medium

  * SnapManager: add threads

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 26 Sep 2018 17:00:44 +0200

lliurex-store (0.5.9) bionic; urgency=medium

  * Add method "policy" to debManager

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 12 Sep 2018 12:30:11 +0200

lliurex-store (0.5.8) bionic; urgency=high

  * Added background xml to mate settings

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Thu, 09 Aug 2018 13:51:05 +0200

lliurex-store (0.5.7.1) bionic; urgency=medium

  * Fix description not loading in appimage plugin

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 16 Jul 2018 12:52:52 +0200

lliurex-store (0.5.7) bionic; urgency=medium

  * Availabe snaps are loaded from available sections and not from wilcard

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 16 Jul 2018 12:21:53 +0200

lliurex-store (0.5.6.5) bionic; urgency=high

  * Reverting to the old new home content

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Wed, 04 Jul 2018 09:20:09 +0200

lliurex-store (0.5.6.4) bionic; urgency=high

  * Changed some home page applications

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Tue, 03 Jul 2018 14:31:44 +0200

lliurex-store (0.5.6.3) bionic; urgency=medium

  * Bug: Store was listing max 10 apps per section

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 03 Jul 2018 12:50:53 +0200

lliurex-store (0.5.6.2) xenial; urgency=medium

  * Minor bugfix. Don't crash when /var/lib/app-info/icons doesn't exists (it must exists but...)

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 03 Jul 2018 10:12:10 +0200

lliurex-store (0.5.6.1) xenial; urgency=medium

  * Re-enable method "list" as it was deactivated for testing.

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 20 Jun 2018 13:44:05 +0200

lliurex-store (0.5.6) xenial; urgency=medium

  * Fixes in various scrolled windows when navigating back and forth
  * Labels in pkg results when searching now have a character limit width

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Wed, 13 Jun 2018 17:34:09 +0200

lliurex-store (0.5.5.1) xenial; urgency=medium

  * Fix typo on storeManage

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 13 Jun 2018 14:13:59 +0200

lliurex-store (0.5.5) xenial; urgency=medium

  * disable debug

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 13 Jun 2018 14:06:36 +0200

lliurex-store (0.5.4.6) xenial; urgency=medium

  * Fix wrong update

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 13 Jun 2018 13:41:28 +0200

lliurex-store (0.5.4.5) xenial; urgency=high

  * Breaks instead of Provides

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Wed, 06 Jun 2018 17:36:45 +0200

lliurex-store (0.5.4.4) xenial; urgency=high

  * Fixing control file

 -- Hector Garcia Huerta <hectorgh@gmail.com>  Wed, 06 Jun 2018 17:11:08 +0200

lliurex-store (0.5.4.3) xenial; urgency=medium

  * Add plugin automanage state status

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 25 Apr 2018 10:01:53 +0200

lliurex-store (0.5.4.2) xenial; urgency=medium

  * Gui now opens snaps
  * Appimages now install in ~/.local/bin

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 18 Apr 2018 14:27:19 +0200

lliurex-store (0.5.4.1) xenial; urgency=medium

  * Add file for external appimages

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 20 Mar 2018 13:38:30 +0100

lliurex-store (0.5.4) xenial; urgency=medium

  * Fix remove method for snaps
  * Add icon cache for snaps
  * Snap plugin enabled by default in lliurex-store-gui

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 20 Mar 2018 12:17:56 +0100

lliurex-store (0.5.3) xenial; urgency=medium

  * Fix debug mode in appimagemanager

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 15 Mar 2018 12:21:16 +0100

lliurex-store (0.5.2) xenial; urgency=medium

  * Add depends on gir1.2-snapd-1>=1.23 and libsnapd-glib>=1.23

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 15 Mar 2018 12:03:10 +0100

lliurex-store (0.5.1) xenial; urgency=medium

  * appImages are now installed on $HOME

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 15 Mar 2018 11:48:01 +0100

lliurex-store (0.5) xenial; urgency=medium

  * Improved argument pass to plugins
  * Add snap support
  * Standarize plugins
  * Reworked storeManager load, search and install methods
  * Added updatable and bundle to infoManager
  * Split bundle support into snap and appimage

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Thu, 25 Jan 2018 16:50:33 +0100

lliurex-store (0.4.2) xenial; urgency=medium

  * Removed text messages from loadstore class

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Fri, 22 Dec 2017 14:06:54 +0100

lliurex-store (0.4.1) xenial; urgency=medium

  * Add python3-bs4 dependency
  * Fix race condition in action "info" when package don't exists

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Wed, 13 Dec 2017 10:08:41 +0100

lliurex-store (0.4-lliurex3) xenial; urgency=medium

  * Fix validation of duplicated desktops in loadStore

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Tue, 07 Nov 2017 11:57:57 +0100

lliurex-store (0.4-lliurex2) xenial; urgency=medium

  * removed text domain from loadStore plugin.

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Mon, 06 Nov 2017 10:45:26 +0100

lliurex-store (0.4-lliurex1) xenial; urgency=medium

  * Removed some debug messages

 -- Juanma Navarro Mañez <juanma1980@gmail.com>  Fri, 20 Oct 2017 09:13:48 +0200

lliurex-store (0.4) xenial; urgency=high


lliurex-store-cli/usr/share/lliurex-store/lliurex-store-cli.py
#!/usr/bin/python3
import sys,os,signal
import argparse
#sys.path.append('/usr/share/lliurex-store')
#sys.path.append('/home/lliurex/lliurex-store/trunk/fuentes/lliurex-appstore.install/usr/share/lliurex-store')
import lliurexstore.storeManager as storeManager
import time
import html2text
import gettext
import threading
gettext.textdomain('python3-lliurex-store')
_=gettext.gettext

class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'

def signal_handler(*args):
	if args[0]==17:
		return
	os.system('setterm -cursor on')
	sys.exit(0)
	

def main():
	def print_results(action=None):
		printed=False
		results=store.get_result(action)
		status=store.get_status(action)
		if not 'status' in status.keys():
			status['status']=1
			status['msg']='package not found'
		processed=[]
		print ("")
		for action in results.keys():
			if action in actions and not actionList[action]:
				if status['status']==0:
					print (_(u"Results for ")+_(action))
					for data in results[action]:
						if action=='info':
							try:
								print(color.DARKCYAN+_(u'Package')+': '+color.END + data['package'])
								print(_(u'Name')+': '+data['name'])
								print(_(u'ID')+': '+data['id'])
								print(_(u'Version')+': '+data['version'])
								print(_(u'Size')+': '+data['size'])
								print(_(u'License')+': '+data['license'])
								listCat=[]
								for cat in data['categories']:
									listCat.append(_(cat))
								print(_(u'Categories')+': '+','.join(listCat))
								msg=''
								if data['state']=='installed':
									msg=_('installed')
								else:
									msg=_('available')
								if data['updatable']:
									msg +=_(' (updatable)')
								print(_(u'Status')+': '+msg)
								print(_(u'Summary')+': '+data['summary'])
								desc=(html2text.html2text(data['description'],"lxml"))
								print(_(u'Description')+': '+desc)
								pkgString=[]
								for dependency in data['depends']:
										pkgName=dependency.split(';')[0]
										pkgString.append(pkgName)
								print(_(u'Depends')+': '+', '.join(pkgString))
								print("")
							except Exception as e:
								print("CLI: Error printing key %s"%e)
						elif action=='search':
							#Only print name and summary
							data_id=''
							printcolor=color.DARKCYAN
							if data['bundle']:
								printcolor=color.PURPLE
							elif (data['package'] not in data['id'] or data['package'] in processed):
								data_id=" (%s)"%data['id']
							else:
								processed.append(data['package'])
							print("%s%s%s%s: %s"%(printcolor,data['package'],data_id,color.END,data['summary']))

						elif action=='list':
							#Print package, summary and status
							try:
								if data['package']:
									package=data['package']
								else:
									package=data['name']
								if data['state']=='installed':
									msg=_('installed')
								else:
									msg=_('available')
								print(color.DARKCYAN+package+color.END+": "+data['summary']+' ('+','.join(data['categories'])+')'+' ('+msg+')')
							except Exception as e:
								print(_(u'Error listing')+ ':'+str(e))
								pass
						elif action=='install':
								print(color.DARKCYAN+data['package']+color.END+" "+ _(u"installed")+" "+color.BOLD+ _(u"succesfully")+color.END)
						elif action=='remove':
								print(color.DARKCYAN+data['package']+color.END+" "+ _(u"removed")+" "+color.BOLD+ _(u"succesfully")+color.END)
						else:
							print("RESULT:\n%s"%data)
				else:
					msg=_(u"Unable to")+' '+_(action)
					failed=parms[action]
					if (action=='search' or action=='info'):
							msg=_(u"Unable to show")
					if action=='list':
							failed=', '.join(failed)

					print (color.RED+_(u"Error")+": "+color.END+msg+' '+failed+' ('+_(status['msg'])+')')
				printed=True
		return(printed)
	#def print_results

	CURSOR_UP='\033[F'
	ERASE_LINE='\033[K'
	actions=[]
	parms={}
	dbg=False
	appimage=False
	snap=False
	autostart=True
	args=process_Args(sys.argv)
#	if args.debug:
#		dbg=True
	if args.appimage:
		appimage=True
	if args.snap:
		snap=True
	if args.update:
		actions.append('cache')
		parms['cache']=None
		autostart=False
	if args.view:
		actions.append('info')
		parms['info']=args.view
	if args.search:
		actions.append('search')
		parms['search']=args.search
	if args.install:
		actions.append('install')
		parms['install']=args.install
	if args.remove:
		actions.append('remove')
		parms['remove']=args.remove
#	if args.random:
#		actions.append('random')
#		parms['random']=args.random
#	if args.list:
#		actions.append('list')
#		parms['list']=args.list

	actionList={'search':False,'info':False,'pkgInfo':False,'install':False,'remove':False,'list':False,'list-sections':False,'random':False,'cache':False}
	start_time=time.time()
	store=storeManager.StoreManager(appimage=appimage,snap=snap,dbg=dbg,autostart=autostart,cli=True)
	for action in actions:
		th=threading.Thread(target=store.execute_action, args = (action,parms[action]))
		th.start()
		actionList[action]=False
		
	inc=0
	banner=' '.join(actions)
	banner='LliureX Store'
	numchar=len(banner)
	os.system('setterm -cursor off')
	while store.is_action_running():
		ini=banner[0:numchar]
		end=banner[numchar:inc]
		text=ini+' '+end
		print(text+'                 ',end='\r')
		numchar-=1
		inc+=1
		time.sleep(0.2)
		if numchar<0:
			numchar=len(banner)
			inc=0
	print("")
	print (CURSOR_UP + ERASE_LINE)
	for key in actionList:
		progressDic=store.get_progress(key)
		if key in progressDic:
				if progressDic[key]==100 and not actionList[key]:
					actionList[key]=print_results(key)
	print_results()
	os.system('setterm -cursor on')

def process_Args(args):
	parser=argparse.ArgumentParser(description=(u'Lliurex Store.'))
	parser.add_argument('-s','--search',metavar='Name',nargs='?',help=(_(u"Search a package")))
	parser.add_argument('-v','--view',metavar='Name',nargs='?',help=(_(u"Show all info from a package")))
	parser.add_argument('-i','--install',metavar='Package',help=(_(u"Install a package")))
	parser.add_argument('-r','--remove',metavar='Package',help=(_(u"Remove a package")))
#	parser.add_argument('--random',metavar='Results',help=(_(u"List random packages")))
#	parser.add_argument('--debug',action='store_true',help=(_(u"Prints debug information")))
	parser.add_argument('--appimage',action='store_true',help=(_(u"Load appimage catalog")))
	parser.add_argument('--snap',action='store_true',help=(_(u"Load snap catalog")))
	parser.add_argument('--update',action='store_true',help=(_(u"Update cache")))
#	parser.add_argument('--list',metavar='list',nargs='?',help=(_(u"List category")))

	args=parser.parse_args()
	return args

#Cause the cli hides the cursor we must assure that the cursor is showing when the program ends
sigs=set(signal.Signals) - {signal.SIGKILL, signal.SIGSTOP}
for sig in sigs:
	signal.signal(sig,signal_handler)
main()
